// admin-bro/lib/backend/controllers/api-controller.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _populator = _interopRequireDefault(require("../utils/populator"));

var _filter = _interopRequireDefault(require("../utils/filter"));

var _viewHelpers = _interopRequireDefault(require("../utils/view-helpers"));

var _forbiddenError = _interopRequireDefault(require("../utils/forbidden-error"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable max-len */

/* eslint no-unused-vars: 0 */
class ApiController {
  /**
   * @param {Object} options
   * @param {AdminBroOptions} options.admin
   * @param {CurrentAdmin} [currentAdmin]
   */
  constructor({
    admin
  }, currentAdmin) {
    this._admin = admin;
    this.currentAdmin = currentAdmin;
  }
  /**
   * Returns context for given action
   * @private
   *
   * @param   {ActionRequest}  request  request object
   * @return  {Promise<ActionContext>} action context
   */


  async getActionContext(request) {
    const {
      resourceId,
      action: actionName
    } = request.params;
    const h = new _viewHelpers.default(this._admin);

    const resource = this._admin.findResource(resourceId);

    const action = resource.decorate().actions[actionName];
    return {
      resource,
      action,
      h,
      currentAdmin: this.currentAdmin,
      _admin: this._admin
    };
  }
  /**
   * Search records by query string.
   *
   * Handler function reponsible for a `.../api/resources/{resourceId}/search/{query}` route
   *
   * @param   {ActionRequest}  request
   *
   * @return  {Promise<SearchResponse>}    found records
   */


  async search(request) {
    const queryString = request.params && request.params.query;

    const resource = this._admin.findResource(request.params.resourceId);

    const decorated = resource.decorate();

    if (!decorated.actions.list.isAccessible(this.currentAdmin, null)) {
      throw new _forbiddenError.default({
        actionName: 'list',
        resourceId: resource.id()
      });
    }

    const titlePropertyName = decorated.titleProperty().name();
    const filters = queryString ? {
      [titlePropertyName]: queryString
    } : {};
    const filter = new _filter.default(filters, resource);
    const records = await resource.find(filter, {
      limit: 50,
      sort: {
        sortBy: titlePropertyName,
        direction: 'asc'
      }
    });
    return {
      records: records.map(record => record.toJSON(this.currentAdmin))
    };
  }
  /**
   * Performs a customized {@link Action resource action}.
   * To call it use {@link ApiClient#resourceAction} method.
   *
   * Handler function reponsible for a `.../api/resources/{resourceId}/actions/{action}`
   *
   * @param   {ActionRequest}  request
   * @param   {Object}  response
   *
   * @return  {Object}  action response
   */


  async resourceAction(request, response) {
    const actionContext = await this.getActionContext(request);

    if (!actionContext.action.isAccessible(this.currentAdmin, null)) {
      throw new _forbiddenError.default({
        actionName: actionContext.action.name,
        resourceId: actionContext.resource.id()
      });
    }

    return actionContext.action.handler(request, response, actionContext);
  }
  /**
   * Performs a customized {@link Action record action}.
   * To call it use {@link ApiClient#recordAction} method.
   *
   * Handler function reponsible for a `.../api/resources/{resourceId}/records/{recordId}/{action}`
   *
   * @param   {ActionRequest}  request
   * @param   {any}  response
   *
   * @return  {Object}  action response
   */


  async recordAction(request, response) {
    const {
      recordId
    } = request.params;
    const actionContext = await this.getActionContext(request);
    let record = await actionContext.resource.findOne(recordId);
    [record] = await (0, _populator.default)([record]);

    if (!actionContext.action.isAccessible(this.currentAdmin, record)) {
      throw new _forbiddenError.default({
        actionName: actionContext.action.name,
        resourceId: actionContext.resource.id()
      });
    }

    return actionContext.action.handler(request, response, { ...actionContext,
      record
    });
  }
  /**
   * Gets optional data needed by the dashboard.
   * To call it use {@link ApiClient#dashboard} method.
   *
   * Handler function reponsible for a `.../api/dashboard`
   *
   * @param   {ActionRequest}  request
   * @param   {any}  response
   *
   * @return  {Promise<any>}  action response
   */


  async dashboard(request, response) {
    const h = new _viewHelpers.default(this._admin);
    const handler = this._admin.options.dashboard && this._admin.options.dashboard.handler;

    if (handler) {
      return handler(request, response, {
        h,
        currentAdmin: this.currentAdmin,
        _admin: this._admin
      });
    }

    return {
      message: 'You can override this method by setting up dashboard.handler fuction in AdminBro options'
    };
  }

}

var _default = ApiController;
/**
 * Response of a Search action in the API
 * @memberof ApiController
 * @alias SearchResponse
 */

exports.default = _default;
