// admin-bro/lib/frontend/components/app/records-table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _recordInList = _interopRequireDefault(require("./record-in-list"));

var _recordsTableHeader = _interopRequireDefault(require("./records-table-header"));

var _noRecords = _interopRequireDefault(require("./no-records"));

var _types = require("../../types");

var _table = _interopRequireDefault(require("../ui/table"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const RecordsTable = props => {
  const {
    resource,
    records,
    actionPerformed,
    sortBy,
    direction,
    isLoading
  } = props;

  if (!records.length) {
    return _react.default.createElement(_noRecords.default, {
      resource: resource
    });
  }

  return _react.default.createElement(_table.default, null, _react.default.createElement(_recordsTableHeader.default, {
    properties: resource.listProperties,
    titleProperty: resource.titleProperty,
    direction: direction,
    sortBy: sortBy
  }), _react.default.createElement("tbody", null, records.map(record => _react.default.createElement(_recordInList.default, {
    record: record,
    resource: resource,
    key: record.id,
    actionPerformed: actionPerformed,
    isLoading: isLoading
  }))));
};

RecordsTable.propTypes = {
  resource: _types.resourceType.isRequired,
  records: _propTypes.default.arrayOf(_types.recordType).isRequired,
  actionPerformed: _propTypes.default.func.isRequired,
  sortBy: _propTypes.default.string,
  direction: _propTypes.default.string.isRequired,
  isLoading: _propTypes.default.bool.isRequired
};
RecordsTable.defaultProps = {
  sortBy: null
};
var _default = RecordsTable;
exports.default = _default;
